### ISSUE TITLE
Metal: autograd weight gradients miss the MPSMatmul rewrites (leading-axis contraction) and fall back to GenericMatmul (~9 vs ~800 GFLOP/s)

### ISSUE THREAD

**Issue by Sheggle** [jafioti/luminal#382]
> [!IMPORTANT]
> **This issue was found and investigated fully autonomously by Opus 5. Feel free to disregard.**
>
> Filed with the repo owner's go-ahead, but the technical content has not been line-by-line reviewed by a human. The analysis and benchmarks below were produced by running the code on real hardware (Apple M2) and the patch was verified against the existing test suite, but please treat the framing with appropriate skepticism and re-verify before acting on it.

## Summary

Forward matmuls lower to `MPSMatmul` and run at full MPS speed, but the weight-gradient matmuls generated by `luminal_training`'s autograd never match the `MPSMatmul` rewrite rules and are extracted as the naive `GenericMatmul` kernel instead. On an M2, that's ~800 GFLOP/s vs ~10-20 GFLOP/s for identical GEMM shapes — in a small conv-net training graph, the two weight-grad GEMMs were ~75% of the whole step's wall-clock.

## Environment

- macOS 15.6 (Darwin 24.6.0), Apple M2, 16 GB
- luminal @ `ab8effa9eaaf356c8635ebfeb50fe51e1ceb33a4` (main, 2026-07-23), `luminal_metal` + `luminal_training` from the same rev
- rustc 1.91.0

## Reproduction

```rust
use luminal::prelude::*;
use luminal_metal::MetalRuntime;
use luminal_training::backward;

const M: usize = 2048;
const N: usize = 2048;
const K: usize = 256;

fn bench(name: &str, f: impl Fn(&mut Graph) -> (Vec<(GraphTensor, usize)>, GraphTensor)) {
    let mut cx = Graph::new();
    let (inputs, out) = f(&mut cx);
    let out = out.output();
    let mut rt = MetalRuntime::initialize(());
    for (t, n) in &inputs {
        rt.set_data(t.id, vec![0.01f32; *n]);
    }
    let mut rt = cx.compile(rt, CompileOptions::default());
    rt.execute(&cx.dyn_map);
    let t0 = std::time::Instant::now();
    let reps = 10;
    for _ in 0..reps {
        for (t, n) in &inputs {
            rt.set_data(t.id, vec![0.01f32; *n]);
        }
        rt.execute(&cx.dyn_map);
        let _ = rt.get_f32(out.id);
    }
    let e = t0.elapsed().as_secs_f64() / reps as f64;
    let flops = 2.0 * M as f64 * N as f64 * K as f64;
    println!("{name}: {:>8.2} ms | {:>6.1} GFLOP/s", e * 1e3, flops / e / 1e9);
}

fn main() {
    // 1. plain forward-shaped GEMM
    bench("fwd-shape    (2048,256)@(256,2048)  ", |cx| {
        let a = cx.tensor((M, K));
        let b = cx.tensor((K, N));
        (vec![(a, M * K), (b, K * N)], a.matmul(b))
    });
    // 2. the same contraction written with an explicit permute — still fast
    bench("permuted-lhs (256,2048)^T@(256,2048)", |cx| {
        let a = cx.tensor((K, M));
        let b = cx.tensor((K, N));
        (vec![(a, M * K), (b, K * N)], a.permute((1, 0)).matmul(b))
    });
    // 3. the same contraction as autograd generates it (weight grad of x@w)
    bench("autograd dW  (realistic loss)       ", |cx| {
        let x = cx.tensor((K, M));
        let w = cx.tensor((M, N));
        let t = cx.tensor((K, N));
        let loss = (x.matmul(w) * t).sum((0, 1));
        let g = backward(cx, loss, &[w])[0];
        (vec![(x, K * M), (w, M * N), (t, K * N)], g)
    });
}
```

Measured (M2):

| case | kernel extracted | time | GFLOP/s |
|---|---|---|---|
| forward-shaped GEMM | `MPSMatmul` | 2.8 ms | ~775 |
| explicit `permute` + matmul | `MPSMatmul` (transposed-lhs rule) | 2.7-4.4 ms | ~490-800 |
| autograd weight grad, same contraction | **`GenericMatmul`** | **227 ms** | **9.4** |

## Root cause

The four `MPSMatmul` rewrite rules (`crates/luminal_metal/src/kernel/ops.rs`, `matmul_rule`) all template the broadcast-mul + sum pair with shape order **(m, n, k)** and the reduction over the **trailing** axis (`iter_stride = z`). That is the form the frontend `matmul` emits, which is why cases 1 and 2 hit MPS.

Autograd, however, emits weight gradients in **batch-leading** form: `dW[m,n] = Σ_k dY[k,m] · X[k,n]`, i.e. mul shape **(k, m, n)** with the reduction over the **leading** axis. `LLIR_DUMP_DIR` for case 3 (with B=256, I=1024, O=512 to make axes unambiguous) shows:

```
GenericMatmul { out_shape: [512, 1024], mul_shape: [256, 512, 1024], k: 256,
                lhs_strides: [(z*512), z, 0],      // dY: (k, m) row-major
                rhs_strides: [(z*1024), 0, z],     // X:  (k, n) row-major
                sum_input_strides: [(z*1024), z],
                sum_iter_stride: (z*524288),       // leading-axis reduction (stride m*n)
                out_strides: [(z*1024), z] }
```

None of the four (m, n, k)/trailing-axis templates match this, so the only rule that fires is the fully-wildcard `GenericMatmul` one, and extraction picks the naive kernel. Semantically the pattern is exactly `MPSMatmul { m, n, k, transpose_lhs: true, transpose_rhs: false }` with the existing stride conventions — only the loop order differs.

## Suggested fix (suggestion only — happy for you to take it or do something better)

Add a second family of four rules, identical to `matmul_rule` but templating shape order **(k, m, n)** with the leading-axis reduction. Two notes on the shape of the rule:

- Per-layout operand strides over (k, m, n) are: lhs stored (k,m) → `[m·z, z, 0]` (transpose_lhs=1) or stored (m,k) → `[z, k·z, 0]`; rhs stored (k,n) → `[n·z, 0, z]` or stored (n,k) → `[z, 0, k·z]` — same `lhs_row_stride`/`rhs_row_stride` conventions as the existing rules.
- The sum's `iter_stride` is `m·n·z`, but templating `mul(mul(m, n), z)` may not match because the product is usually a pre-simplified literal in the e-graph. Binding it structurally avoids that: pattern the mul's `out_strides` as `(ICons ?it (ICons ?om (ICons ?on (INil))))` and require the sum to read with `strides = [?om, ?on]`, `iter_stride = ?it`. That ties the reduction to the mul's leading output axis without any expression algebra.

A verified reference implementation (~130 added lines, mirroring the existing `matmul_rule` closure style) is attached as a patch below. With it, on the same machine:

- case 3 above: **227 ms → 9.6 ms** (9.4 → 225 GFLOP/s, ~24x),
- a real training step (2 locally-connected GEMM layers + trunk + head, batch 256, fwd+bwd+AdamW in one graph): **312 ms → 77 ms** (4.1×),
- all 45 existing `luminal_metal` tests and the `luminal_training` tests pass, and gradients match the reference runtime bit-for-bit on the graphs checked.

Known remaining gap (fine to ignore): when the upstream gradient is a broadcast **constant** (e.g. `loss = y.sum()`, so dY has strides `[0,0,0]`), none of the layout templates match and it still falls back to `GenericMatmul`. That case is degenerate (the matmul is really a rank-1 sum) and doesn't appear with real losses.

<details>
<summary>patch (reference implementation)</summary>

```diff
diff --git a/crates/luminal_metal/src/kernel/ops.rs b/crates/luminal_metal/src/kernel/ops.rs
index a3c0edd..3e2ae4e 100644
--- a/crates/luminal_metal/src/kernel/ops.rs
+++ b/crates/luminal_metal/src/kernel/ops.rs
@@ -1420,7 +1420,137 @@ impl EgglogOp for MPSMatmul {
                 .name(name)
         };
 
+        // Same rewrite family for matmuls whose contraction axis LEADS the
+        // broadcast-mul shape — (k, m, n) reduced over axis 0 — which is the
+        // form autograd emits for weight gradients (dW[m,n] = sum_k
+        // lhs[k,m] * rhs[k,n]). The sum's iter_stride is tied to the mul's
+        // leading output stride by pattern instead of templating m*n*z, which
+        // may not exist as a decomposed expression node.
+        let matmul_rule_leading_k = |name: &'static str,
+                                     lhs_layout: MPSMatrixLayout,
+                                     rhs_layout: MPSMatrixLayout,
+                                     transpose_lhs: i64,
+                                     transpose_rhs: i64| {
+            let m = v("?m");
+            let n = v("?n");
+            let k = v("?k");
+            let lhs = v("?lhs");
+            let rhs = v("?rhs");
+            let lhs_row_stride = match lhs_layout {
+                MPSMatrixLayout::RowMajor => mul(k.clone(), z.clone()),
+                MPSMatrixLayout::TransposedRowMajor => mul(m.clone(), z.clone()),
+            };
+            let rhs_row_stride = match rhs_layout {
+                MPSMatrixLayout::RowMajor => mul(n.clone(), z.clone()),
+                MPSMatrixLayout::TransposedRowMajor => mul(k.clone(), z.clone()),
+            };
+            // Axis order (k, m, n).
+            let lhs_strides = match lhs_layout {
+                // stored (m, k) row-major
+                MPSMatrixLayout::RowMajor => vec![z.clone(), lhs_row_stride.clone(), zero.clone()],
+                // stored (k, m) row-major
+                MPSMatrixLayout::TransposedRowMajor => {
+                    vec![lhs_row_stride.clone(), z.clone(), zero.clone()]
+                }
+            };
+            let rhs_strides = match rhs_layout {
+                // stored (k, n) row-major
+                MPSMatrixLayout::RowMajor => vec![rhs_row_stride.clone(), zero.clone(), z.clone()],
+                // stored (n, k) row-major
+                MPSMatrixLayout::TransposedRowMajor => {
+                    vec![z.clone(), zero.clone(), rhs_row_stride.clone()]
+                }
+            };
+            let out_row_stride = mul(n.clone(), z.clone());
+            let iter_stride = v("?gk_iter_stride");
+            let om = v("?gk_om");
+            let on = v("?gk_on");
+
+            let mul_op = op_term(
+                MetalMul {
+                    shape: vec![],
+                    a_strides: vec![],
+                    b_strides: vec![],
+                    output_strides: vec![],
+                }
+                .sort()
+                .call([
+                    (
+                        "shape",
+                        cons(k.clone(), cons(m.clone(), cons(n.clone(), nil()))),
+                    ),
+                    ("a_strides", expr_list(lhs_strides)),
+                    ("b_strides", expr_list(rhs_strides)),
+                    (
+                        "out_strides",
+                        cons(iter_stride.clone(), cons(om.clone(), cons(on.clone(), nil()))),
+                    ),
+                ]),
+                ilist(vec![lhs.clone(), rhs.clone()]),
+            );
+            let sum_op = op_term(
+                MetalSumReduce::default().sort().call([
+                    ("shape", cons(m.clone(), cons(n.clone(), nil()))),
+                    ("iters", k.clone()),
+                    ("strides", cons(om.clone(), cons(on.clone(), nil()))),
+                    ("iter_stride", iter_stride.clone()),
+                    (
+                        "out_strides",
+                        cons(out_row_stride.clone(), cons(z.clone(), nil())),
+                    ),
+                ]),
+                ilist(vec![mul_op.clone()]),
+            );
+            let mps_op = MPSMatmul::default().sort().call([
+                ("m", m),
+                ("n", n),
+                ("k", k),
+                ("lhs", lhs),
+                ("lhs_row_stride", lhs_row_stride),
+                ("rhs", rhs),
+                ("rhs_row_stride", rhs_row_stride),
+                ("out_row_stride", out_row_stride),
+                ("transpose_lhs", lit_i64(transpose_lhs)),
+                ("transpose_rhs", lit_i64(transpose_rhs)),
+            ]);
+            let dt = v(format!("?{}_dt", name.replace('-', "_")));
+
+            rule(union(sum_op.clone(), mps_op.clone()))
+                .set(dtype(mps_op), dt.clone())
+                .fact(eq(dt, dtype(sum_op)))
+                .ruleset("kernel_lower")
+                .name(name)
+        };
+
         vec![
+            matmul_rule_leading_k(
+                "mps-matmul-leading-k-row-row",
+                MPSMatrixLayout::RowMajor,
+                MPSMatrixLayout::RowMajor,
+                0,
+                0,
+            ),
+            matmul_rule_leading_k(
+                "mps-matmul-leading-k-row-transposed-rhs",
+                MPSMatrixLayout::RowMajor,
+                MPSMatrixLayout::TransposedRowMajor,
+                0,
+                1,
+            ),
+            matmul_rule_leading_k(
+                "mps-matmul-leading-k-transposed-lhs-row",
+                MPSMatrixLayout::TransposedRowMajor,
+                MPSMatrixLayout::RowMajor,
+                1,
+                0,
+            ),
+            matmul_rule_leading_k(
+                "mps-matmul-leading-k-transposed-lhs-transposed-rhs",
+                MPSMatrixLayout::TransposedRowMajor,
+                MPSMatrixLayout::TransposedRowMajor,
+                1,
+                1,
+            ),
             matmul_rule(
                 "mps-matmul-row-row",
                 MPSMatrixLayout::RowMajor,
```

</details>

---

Context: found while building a small AlphaZero training loop on the Metal backend, shortly after #374 landed. Given the PR notes that the `Trainer` is currently special-cased to the `ReferenceBackend`, this is presumably just a path that has not been exercised yet rather than a regression. Developed against `ab8effa`; `crates/luminal_metal/src/kernel/ops.rs` appears untouched by `d4a689f`, so the patch should still apply.
